Blockchain API and Mosaic API

vulfke482 opened the pull request:
Blockchain API: All queries are done, except getting block's transactions, because it needs to implement transaction deserialization.
Mosaic API: Done.

vulfke482 committed:
Added custom json serialization and translated messages

vulfke482 committed:
Updated variadic_struct.h from develop

vulfke482 committed:
Remade serialization via custom serialization method

vulfke482 committed:
backup

vulfke482 committed:
second try to save commit while uint64 doesn't work

vulfke482 committed:
It is a half a way for masaic, it is compilling anyway

vulfke482 committed:
Ended skeleton for mosaics

vulfke482 committed:
All review notices, except dto architecture and build artefacts

vulfke482 committed:
Mosaics are finished

vulfke482 committed:
Resolved problems with DTO

vulfke482 committed:
Added methods for chain score and storage info requests.

vulfke482 committed:
Changes in masaics and mesasges after review

vulfke482 committed:
Debuging problems with server

vulfke482 committed:
Updates:
Implemented namespace api infrastructure.
Implemented account api infrastructure.
Resolved noticed in review problems.
Deleted lower_higher.h/.cpp
Added exception processing.

vulfke482 committed:
Minor changes in mosaic service

vulfke482 committed:
Changed input parameters in masaic api requests.

vulfke482 committed:
Added virtual destructor to messages.h:Message.

vulfke482 committed:
Namespace queries are done. Beginning of refactoring.

vulfke482 committed:
Refactoring is complited. Cannot link 'lib/libnemcpp.so.1.0.0'

vulfke482 committed:
Refactoring is done.

vulfke482 committed:
Added account and network services

vulfke482 committed:
Added general transaction dto.

vulfke482 committed:
Transactions: draft

otsybizov commented on include/nemcpp/client.h:
```diff
@@ -0,0 +1,37 @@
+		static std::string getErrorMessage(uint16_t code);
+	};
+
+	class InvalidJSON : public std::runtime_error {
```
Please, use camel case. E.g. InvalidJSON -> InvalidJson

vulfke482 replied:
Done

otsybizov commented on include/nemcpp/contract/ContractService.h:
```diff
@@ -0,0 +1,33 @@
+*** license that can be found in the LICENSE file.
+*/
+
+#ifndef NEM2_SDK_CPP_CONTRACTSERVICE_H
```
Please, use #pragma once instead of inclusion guard.

vulfke482 replied:
I have deleted whole file :)

otsybizov commented on include/nemcpp/model/account/account_property.h:
```diff
@@ -24,38 +37,38 @@ namespace nem2_sdk {
 	template<typename TValue, AccountPropertyType PropertyTypeId>
 	struct AccountProperty {
```
Class with the same name is already defined above, that's a bit confusing.

vulfke482 replied:
They are in different namespaces. Nothing to do with that for now.

otsybizov commented on include/nemcpp/model/account/address.h:
```diff
@@ -88,10 +88,10 @@ namespace nem2_sdk {
 namespace std {
```
It is not recommended to extend std namespace.

pananton replied:
C++ standard specifies that it is allowed to extend std namespace _specializing_ existing templates for _user-defined_ types (like std::hash for example, or std::tuple_size). Maybe it's a bad practice for some reason, but I haven't heard about this.

otsybizov commented on include/nemcpp/model/message/message.h:
```diff
@@ -0,0 +1,175 @@
+        }
+
+    private:
+        ByteArray payload; //ByteArray = struct { uint8_t* array, int size;}
```
Please, use m_ prefix for class member names.

pananton replied:
Please note, that m_ prefix is not used in existing library types. If you want m_ prefix better to rename members of other classes for consistency (currently data members are named with "_" suffix for private data members and as-is for public struct's data members).

otsybizov commented on include/nemcpp/service/http.h:
```diff
@@ -0,0 +1,26 @@
+#include <boost/asio/ip/tcp.hpp>
+
+namespace xpx_sdk::http {
+	typedef int StatusCode;
```
Prefer 'using' to 'typedef'.

vulfke482 replied:
Done

otsybizov commented on src/sdk/model/mosaic/mosaic_info.cpp:
```diff
@@ -0,0 +1,9 @@
+*** license that can be found in the LICENSE file.
+*/
+
+class MosaicSerivice {
```
What is it for?

vulfke482 replied:
This file is no longer with us too)

otsybizov commented:
Overall, I don't see any benefit of interfaces like IClient, IAccountService etc. You'd better remove them at all.

vulfke482 committed:
backup

vulfke482 committed:
backup2

vulfke482 committed:
Trying to resolve problems with binary serialization

vulfke482 committed:
before deleting all binary serialization

vulfke482 committed:
Almost builded

vulfke482 committed:
Almost build 2

vulfke482 committed:
Working version

vulfke482 committed:
GET and POST requests are working

vulfke482 committed:
Added optional fields in DTO

eugene-babichenko committed:
fix build on macOS

eugene-babichenko commented on include/nemcpp/model/transaction/account_property_transaction_types.h:
```diff
@@ -8,7 +8,7 @@
 #include <set>
 
-namespace nem2_sdk {
+namespace xpx_sdk::difficult_transactions {
```
maybe `complex_transactions`?

vulfke482 replied:
Changed

eugene-babichenko committed:
Merge pull request #13 from proximax-storage/feature/http-websockets-impl

Implement HTTP connection & add a sample request implementation

vulfke482 committed:
Added several overloaded methods

vulfke482 committed:
Rebase is done, but there are some problems with connection.

vulfke482 committed:
Removed cmake-build-debug

eugene-babichenko commented on .gitignore:
```diff
@@ -1,2 +1,3 @@
 _build/
```
Please remove.

vulfke482 replied:
Removed

eugene-babichenko commented on .gitignore:
```diff
@@ -1,2 +1,3 @@
-build/
\ No newline at end of file
+build/
+cmake-build-debug/
```
Add a newline

vulfke482 replied:
Added

eugene-babichenko commented on CMakeLists.txt:
```diff
@@ -196,7 +198,7 @@ if (NEM2_SDK_BUILD_TESTS)
 
 if (NEM2_SDK_BUILD_EXAMPLES)
-	add_executable(get_block ${CMAKE_CURRENT_SOURCE_DIR}/examples/get_block.cpp)
+	add_executable(get_block ${CMAKE_CURRENT_SOURCE_DIR}/examples/get_block.cpp )
```
Remove whitespace in the end of the line

vulfke482 replied:
Done

eugene-babichenko commented on include/nemcpp/client/namespace_service.h:
```diff
@@ -0,0 +1,36 @@
+        virtual MultipleNamespaceInfo getNamespaceInfoByAccounts(const std::vector<std::string>& accountIds) = 0;
+        virtual NamespaceNames getNamespaceNames(const std::vector<std::string>& namespaceIds) = 0;
+		virtual NamespaceInfo getNamespaceInfoById(const std::string& id) = 0;
+//        virtual std::vector<string> getAccountIds(); //maybe in future if will know how to do it
```
Commits to be merged must not contain commented lines.

eugene-babichenko commented on include/nemcpp/contract/ContractService.h:
```diff
@@ -0,0 +1,33 @@
+#ifndef NEM2_SDK_CPP_CONTRACTSERVICE_H
+#define NEM2_SDK_CPP_CONTRACTSERVICE_H
+
+//#include "context.h"
```
Please remove commented lines or finish working on them.

vulfke482 replied:
Removed whole file)

eugene-babichenko commented on include/nemcpp/model/account/account.h:
```diff
@@ -70,7 +72,7 @@ namespace nem2_sdk {
 		
 		/// Calculates cosignature for aggregate transaction.
-		Signature cosignAggregateTransaction(const AggregateTransaction* transaction) const;
+//		Signature cosignAggregateTransaction(const AggregateTransaction* transaction) const;
```
Please remove commented lines or finish working on them.

eugene-babichenko commented on include/nemcpp/model/transaction/secret_lock_transaction.h:
```diff
@@ -64,49 +64,49 @@ namespace nem2_sdk {
-	                                              const Address& lockedMosaicRecipient,
-	                                              const Key& signer,
-	                                              std::optional<NetworkIdentifier> networkId = std::nullopt);
+//	/// Creates secret lock transaction.
```
Please remove commented lines or finish working on them.

eugene-babichenko commented on include/nemcpp/model/transaction_simple/aggregate_transaction_types.h:
```diff
@@ -0,0 +1,16 @@
+namespace xpx_sdk::simple_transactions {
+	
+	/// Transactions, embedded to aggregate transaction.
+//	using EmbeddedTransactions = std::vector<std::shared_ptr<EmbeddedTransaction>>;
```
Please remove commented lines or finish working on them.

eugene-babichenko commented on src/infrastructure/binary/dto/transaction_dto.h:
```diff
@@ -54,9 +56,9 @@ namespace nem2_sdk { namespace internal { namespace binary {
-	using AddressPropertyModificationDTO = TAccountPropertyModificationDTO<AccountAddressProperty::ValueType>;
-	using MosaicPropertyModificationDTO = TAccountPropertyModificationDTO<AccountMosaicProperty::ValueType>;
-	using TransactionPropertyModificationDTO = TAccountPropertyModificationDTO<AccountTransactionProperty::ValueType>;
+//	using AddressPropertyModificationDTO = TAccountPropertyModificationDTO<AccountAddressProperty::ValueType>;
```
Please remove commented lines or finish working on them.

eugene-babichenko commented on src/infrastructure/network/http.cpp:
```diff
@@ -44,9 +46,13 @@ std::string _performHTTPRequest_internal(
 	if (method == HTTPRequestMethod::PUT || method == HTTPRequestMethod::POST) {
-		request.set(http::field::body, request_body);
+		request.set(http::field::content_length, request_body.size());
+//		request.set(http::field::body, request_body);
```
Please remove commented lines or finish working on them.

eugene-babichenko commented on src/infrastructure/network/http.h:
```diff
@@ -4,12 +4,12 @@
 		GET,
 		POST,
-		PUT,
+		PUT
```
Please add comma in the end. This will provide us clearer diffs in the future

eugene-babichenko commented on src/infrastructure/network/websocket.cpp:
```diff
@@ -1,123 +1,119 @@
-}
-
-#endif
+//#include "websocket.h"
```
Revert this change.

eugene-babichenko commented on src/infrastructure/network/websocket.h:
```diff
@@ -1,78 +1,75 @@
-#ifdef NEM2_SDK_ENABLE_WEBSOCKETS
+//#pragma once
```
Revert this change.

eugene-babichenko commented on src/sdk/client/mosaic_service.cpp:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    Parser::Write(hexIds, requestJson);
+//	std::string tmp;
```
Please remove commented lines or finish working on them.

vulfke482 replied:
Done

eugene-babichenko commented on src/sdk/client/namespace_service.cpp:
```diff
@@ -0,0 +1,104 @@
+NamespaceNames NamespaceService::getNamespaceNames(const std::vector<std::string> & namespaceIds) {
+    std::string requestJson;
+    Parser::Write(namespaceIds, requestJson);
+    std::stringstream path;
```
Just use strings if the request URL is static. Please also look at other places using using `std::stringstream`.

vulfke482 replied:
Done for all POST requests. GET requests without parameters have been already done in this way.

eugene-babichenko commented on src/sdk/model/account/account.cpp:
```diff
@@ -55,17 +55,17 @@ namespace nem2_sdk {
-		KeyPair keyPair(requestPrivateKey(PrivateKeySupplierReason::Aggregate_Transaction_Cosigning, transaction));
-		return CalculateAggregateTransactionCosignature(transaction, keyPair);
-	}
+//	Signature Account::cosignAggregateTransaction(const AggregateTransaction* transaction) const
```
Please remove commented lines or finish working on them.

vulfke482 replied:
Removed part was moved to other branch.

eugene-babichenko commented on src/sdk/model/transaction/aggregate_transaction.cpp:
```diff
@@ -1,32 +1,32 @@
-namespace nem2_sdk {
-	
+namespace xpx_sdk::difficult_transactions {
+//
```
what is this for?

vulfke482 replied:
Removed

eugene-babichenko commented on src/sdk/model/transaction/create_transaction.cpp:
```diff
@@ -10,11 +10,16 @@
-using namespace nem2_sdk::internal::binary;
+using namespace xpx_sdk::internal;
+using namespace xpx_sdk::internal::binary;
+//using namespace xpx_sdk::difficult_transactions;
```
Please remove commented lines or finish working on them.

vulfke482 replied:
removed

eugene-babichenko reviewed (changes requested):
This pull request is confirmed to be working, so I am not looking deep into the code (I believe that the rest of the team have done so in the previous two weeks). Most of my comments are regarding the code style.

In addition to the code comments I would like to outline the following:

- Please remove the commented pieces of code. If you still need to work on them, move them to a separate branch to make this one clean.
- Provide descriptions for your TODO comments.
- Address pending reviews from this pull request as well as from #7. Implement them, or respond to them, or close the discussions so that I can see there is an agreement.

After that I will consider this pull request ready to merge.

I do not concentrate on the code formatting here. This will be done automatically in a separate pull request.

vulfke482 committed:
Changes in services after review

vulfke482 committed:
Changes after review

vulfke482 committed:
changed InvalidJSON to camel case

vulfke482 committed:
Version after review

vulfke482 committed:
Beginning of brutforce

vulfke482 committed:
Deleted commented code in read_json.h

vulfke482 committed:
All DTO convertions added.

vulfke482 committed:
Transaction deserialization finished

vulfke482 committed:
Blockchain and Account transaction queries created

vulfke482 committed:
Transactions added

vulfke482 committed:
Resolving conflicts

eugene-babichenko commented:
From d596666 on macOS

```
7697208349161298311
Mosaic Info: 9000000000000000
1
libc++abi.dylib: terminating with uncaught exception of type xpx_sdk::serialization_error: Exception: Cannot parse JSON. Error with: (/transaction/size)
[1]    3415 abort      ./bin/get_block
```

vulfke482 committed:
Transactions are fixed

vulfke482 committed:
Merge branch 'transactions-t52' into feature/api-skeleton

vulfke482 committed:
Added conversion to dto for transactions

vulfke482 committed:
Fixed convertion from mosaic to DTO

vulfke482 committed:
Removed interfaces for sdk services.

vulfke482 committed:
Fixed compilation error

vulfke482 committed:
Corrected aggregate transaction, still needs reconstruction.

vulfke482 committed:
Merge branch 'transactions-t52' into feature/api-skeleton

Added conversion from transacton to DTO.
Removed interface for every service.

vulfke482 committed:
Transaction service 4 out of 7

vulfke482 committed:
Before I rename half of my namespaces

vulfke482 committed:
Transaction sending is done, testing.

vulfke482 committed:
Transaction sending works. Checking binary serialization.

vulfke482 committed:
Transaction sending is done.

vulfke482 committed:
Updates on transactions:

Created new namespace for information objects;
Now they are demarcated from transaction creation objects and functions;
Transaction sending is working, serialization behaves well;

vulfke482 committed:
Merge branch 'transactions-t52' into feature/api-skeleton

Transactions sending is working;
Server haven't accepted any transaction yet, due to format issues;

vulfke482 committed:
Removed unused code.

vulfke482 committed:
Sended transaction is accepted by server.
